chore(migrate): Use the migration engine JSON-RPC printing API

tomhoule opened the pull request:
https://github.com/prisma/prisma-engines/pull/2566
todo: explain better

Note:
https://www.jsonrpc.org/specification

tomhoule committed:
chore(migrate): Use the migration engine JSON-RPC printing API

todo: explain better

tomhoule commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -139,15 +140,11 @@ export class MigrateEngine {
           this.listeners[result.id](result)
           delete this.listeners[result.id]
         }
-      } else {
```
Deleted because I'm confident this code path is never taken, based on what the ME returns.

Jolg42 commented:
@tomhoule Let's sync about this tomorrow?

tomhoule commented:
yep, good idea, let's do that :+1:

Jolg42 commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -136,6 +136,7 @@ export class MigrateEngine {
     }
+
     if (result) {
       if (result.id) {
```
```suggestion
      // It's a response
      if (result.id) {
```

Jolg42 commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -145,15 +146,11 @@ export class MigrateEngine {
-        }
+      } else if (result.method) {
+          // This is a notification.
+          if (result.method === "print" && Boolean(result.params?.content)) {
```
```suggestion
          if (result.method === "print" && result.params?.content) {
```

Jolg42 commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -145,15 +146,11 @@ export class MigrateEngine {
+      } else if (result.method) {
+          // This is a notification.
+          if (result.method === "print" && Boolean(result.params?.content)) {
+              process.stdout.write(result.params?.content)
```
```suggestion
              process.stdout.write(result.params.content)
```

tomhoule committed:
Apply suggestions from code review

Co-authored-by: Joël Galeran <Jolg42@users.noreply.github.com>

Jolg42 committed:
update engine

3.10.0-13.migration-engine-json-rpc-print-d0b4704320a9d5a5e3e86e2b8db562ac9137aab3

Jolg42 committed:
use console.info for tests (console.info is mocked)

Jolg42 committed:
wait for engine notification(s), tweak line returns, update snapshots

Jolg42 commented on packages/migrate/src/__tests__/MigrateDev.test.ts:
```diff
@@ -435,6 +435,7 @@ describe('sqlite', () => {
       The migration \`20201231000000_test\` was modified after it was applied.
 
       Applying migration \`20201231000000_test\`
+
       Applying migration \`20201231000000_draft\`
```
@tomhoule this doesn't look good

Jolg42 committed:
better order of logs for one migrate dev case

Jolg42 committed:
update engines

Jolg42 committed:
put empty lines back

Jolg42 committed:
Merge remote-tracking branch 'origin/main' into migrate/json-rpc-print

# Conflicts:
#	packages/cli/package.json
#	packages/client/package.json
#	packages/engine-core/package.json
#	packages/migrate/package.json
#	packages/migrate/src/__tests__/MigrateDiff.test.ts
#	packages/sdk/package.json
#	pnpm-lock.yaml

Jolg42 committed:
use latest engine

Jolg42 committed:
snapshot update

Jolg42 committed:
update diff snapshots

Jolg42 committed:
update 3.10.0-25.bbfe2ad95a483f6374e0f6fa70a2d00861a08703

Jolg42 committed:
update snapshot

Jolg42 commented on packages/migrate/src/commands/MigrateDiff.ts:
```diff
@@ -244,6 +244,13 @@ ${chalk.bold('Examples')}
         shadowDatabaseUrl: args['--shadow-database-url'],
       })
     } finally {
+      // We need to wait for the "notification(s)" sent by the engine via JSON-RPC to be printed to console.info
+      // In Jest it manifests also as the following if we don't wait:
+      // Cannot log after tests are done. Did you forget to wait for something async in your test?
+      //
+      // we need to return the value in the response then we can remove this
+      await new Promise((resolve) => setTimeout(resolve, 50))
```
@tomhoule we really need to clean this up soon and pass the diff in the response 🙈

tomhoule replied:
I am pretty sure asynchronously waiting here is not necessary. What we did before is send the notifications and the response on different channels (stderr and stdout), but with the companion PR that is already merged, the print notification is guaranteed to come before the response on stdout, so if the ordering of operations in the CLI matches the ordering of the JSON-RPC messages on stdout, there is no need for extra logic.

If we want to impose more synchronization and delays (e.g. if the log call and the response are reordered by the event loop), we could make the notification a request — the CLI would then send a response (an ACK), and the engine would wait for that before sending its own response. Long term, I think we want to use this mechanism a lot more, not less, because it avoids spreading control flow across both sides of the RPC boundary.

millsp replied:
Thanks, that makes a lot of sense. Hopefully, this will resolve most of the flakyness we see today :100:

Jolg42 replied:
@tomhoule the tests fail when we do not wait, see
https://github.com/prisma/prisma/runs/5182960010?check_suite_focus=true

Jolg42 committed:
wait for `Applying migration ....`

tomhoule commented on packages/migrate/src/__tests__/MigrateDev.test.ts:
```diff
@@ -629,14 +629,14 @@ describe('sqlite', () => {
     await expect(result).rejects.toMatchInlineSnapshot(`
 
-                                                                                                                                                                                                                                                                                                                                                                                                                ...
+                                                                                                                                                                                                                                                                                                                                                                                                                ...
```
Is this expected? I'm a bit puzzled

Jolg42 replied:
Jest snapshots do that for a reason we don't completely understand.
You can ignore

tomhoule commented on packages/migrate/src/commands/MigrateDeploy.ts:
```diff
@@ -121,6 +121,9 @@ ${editedMigrationNames.join('\n')}`,
     } finally {
+      // We need to wait for the "notification(s)" sent by the engine via JSON-RPC to be printed to console.info
+      await new Promise((resolve) => setTimeout(resolve, 50))
+      // Stop engine
```
(see other comment, but I think it's the same here)

tomhoule commented on packages/migrate/src/commands/MigrateDev.ts:
```diff
@@ -175,6 +175,9 @@ ${chalk.bold('Examples')}
       migrationIdsApplied.push(...appliedMigrationNames)
 
+      // We need to wait for the "notification(s)" sent by the engine via JSON-RPC to be printed to console.info
+      await new Promise((resolve) => setTimeout(resolve, 50))
```
same here

tomhoule reviewed (commented):
Since these waits are not nice at all, let's find a solid solution instead. Can you confirm they are really needed with an up-to-date version of engines?

Jolg42 committed:
comment setTimeout

Jolg42 committed:
Revert "comment setTimeout"

This reverts commit a4ee78e4378ee85e9feb200ecb7cdfc42cf99cd2.

Jolg42 reviewed (approved):
Thanks for the comments,
Indeed, we can improve the communication of CLI - engine here.

I will merge this now to unblock the pipeline.
